pkg/raft/rawnode.go

```diff
@@ -298,51 +297,19 @@ func newStorageAppendRespMsg(r *raft, rd Ready) pb.Message {
-		// MsgStorageAppendResp if returned, the response is ignored and the unstable
-		// log is not truncated. The unstable log is only truncated when the term has
-		// remained unchanged from the time that the MsgStorageAppend was sent to the
-		// time that the MsgStorageAppendResp is received, indicating that no-one else
-		// is in the process of truncating the stable log.
+		// If accTerm has changed by the time the MsgStorageAppendResp is returned,
+		// the response is ignored and the unstable log is not truncated. The
+		// unstable log is only truncated when the term has remained unchanged from
+		// the time that the MsgStorageAppend was sent to the time that the response
+		// is received, indicating that no new leader has overwritten the log.
 		//
-		// However, this replaces a correctness problem with a liveness problem. If we
-		// only attempted to truncate the unstable log when appending new entries but
-		// also occasionally dropped these responses, then quiescence of new log entries
-		// could lead to the unstable log never being truncated.
-		//
-		// To combat this, we attempt to truncate the log on all MsgStorageAppendResp
-		// messages where the unstable log is not empty, not just those associated with
-		// entry appends. This includes MsgStorageAppendResp messages associated with an
-		// updated HardState, which occur after a term change.
-		//
-		// In other words, we set Index and LogTerm in a block that looks like:
-		//
-		//  if r.raftLog.hasNextOrInProgressUnstableEnts() { ... }
-		//
-		// not like:
-		//
-		//  if len(rd.Entries) > 0 { ... }
-		//
-		// To do so, we attach r.raftLog.lastIndex() and r.raftLog.lastTerm(), not the
-		// (index, term) of the last entry in rd.Entries. If rd.Entries is not empty,
-		// these will be the same. However, if rd.Entries is empty, we still want to
-		// attest that this (index, term) is correct at the current term, in case the
-		// MsgStorageAppend that contained the last entry in the unstable slice carried
-		// an earlier term and was dropped.
-		//
-		// A MsgStorageAppend with a new term is emitted on each term change. This is
-		// the same condition that causes MsgStorageAppendResp messages with earlier
-		// terms to be ignored. As a result, we are guaranteed that, assuming a bounded
-		// number of term changes, there will eventually be a MsgStorageAppendResp
-		// message that is not ignored. This means that entries in the unstable log
-		// which have been appended to stable storage will eventually be truncated and
-		// dropped from memory.
+		// TODO(pav-kv): unstable entries can be partially released even if the last
+		// accepted term changed, if we track the (term, index) points at which the
+		// log was truncated.
 		//
 		// [^1]: https://en.wikipedia.org/wiki/ABA_problem
-		last := r.raftLog.lastEntryID()
-		m.Index = last.index
-		m.LogTerm = last.term
+		m.Index = rd.Entries[ln-1].Index
+		m.LogTerm = r.raftLog.accTerm()
```

Comment:
@nvanbenschoten Could use `m.Term` field instead of `m.LogTerm`, if we don't want to overload the meaning of `m.LogTerm`. In `MsgApp` messages, `m.LogTerm` is the "entry term", while in this protocol it carries the "last accepted / leader term".

Maybe using `m.Term` for this would be more semantically consistent - `m.Term` is usually the leader sending a `MsgApp`, and `MsgStorageAppend` is essentially `MsgApp` forwarded to the storage on behalf of the latest leader with whom our log is consistent.

WDYT?